src/DynamoCoreWpf/ViewModels/Core/AnnotationViewModel.cs

```diff
@@ -193,7 +512,149 @@ public AnnotationViewModel(WorkspaceViewModel workspaceViewModel, AnnotationMode
                 var selectNothing = new DynamoModel.SelectModelCommand(Guid.Empty, System.Windows.Input.ModifierKeys.None.AsDynamoType());
                 WorkspaceViewModel.DynamoViewModel.ExecuteCommand(selectNothing);
             }
-            
+
+            InPorts = new ObservableCollection<PortViewModel>();
+            OutPorts = new ObservableCollection<PortViewModel>();
+            InputNodes = new ObservableCollection<NodeViewModel>();
+            OutputNodes = new ObservableCollection<NodeViewModel>();
+
+            InPorts.CollectionChanged += InPorts_CollectionChanged;
+            OutPorts.CollectionChanged += OutPorts_CollectionChanged;
+
+            ViewModelBases = this.WorkspaceViewModel.GetViewModelsInternal(annotationModel.Nodes.Select(x => x.GUID));
+            IsExpanded = annotationModel.IsExpanded;
+
+            // Add all grouped AnnotaionModels to the CutGeometryDictionary.
+            ViewModelBases.OfType<AnnotationViewModel>()
+                .ToList()
+                .ForEach(x => AddToCutGeometryDictionary(x));
+        }
+
+        /// <summary>
+        /// Creates input ports for the group based on its Nodes.
+        /// Input ports that either is connected to a Node outside of the
+        /// group, or has a port that is not connected will be used for the group.
+        /// </summary>
+        internal void SetGroupInputPorts()
```

Comment:
Would you add some comments in this function body? I think it is quite readable but I want to make sure other team member know what is the process